### DISCUSSION TITLE
Questions / help request related to SCRIPT

### DISCUSSION THREAD

**Discussion by Pewibe** [arendst/Tasmota#18433] [Q&A] [answered]
After writing a larger script and a couple of coding trials there are some items still open:
1. Because I need variables 32 bit range I used the 4 counters. Is it possible to change their name per script? Or is it possible to suppress their appearance in web interface?
2. In script examples I found text sequences like {m} for a tab. Where to be find a list of such codes?
3. The script section >T is not in sync with the teleperiod produced by sensors (like counters mentioned above). Any chance to synchronize both or suppress the 'other' teleperiod cycle? What about a feature request?
4. Error: Sektion '>WM' is not recognized (treated as '>W').

**Comment by pkkrusty** [comment]
@gemu2015

**Comment by gemu2015** [comment]
1. what do you want to achieve ? while tasmota counters are indeed 32 bit in scripting they are handled as single float which is 23 bit precision.
if you really need 32 bit precision you have 2 options:
 a. recompile with #define TS_FLOAT double. now all math is double precision. this however uses double RAM memory and is much slower
 b. define a variable as "integer"  I:var=0,  there are various limitations but this should solve most problems with limited precision
2. {m} is a sequence that is replaced by tasmota javascript to a TAB.  while there are several of these i am not aware of a list of such sequences.
3.  >T is called at least at tele period time, however sensors are scanned much faster, up to every 100 ms.
4. >WM must appear before >W in script to be recognized.

**Reply by Pewibe** [reply]
Thanks Gerhard, your answers were quick but not sufficient. 
1. My 2 questions were not answered, Any hope for me?
1a. Too much negative aspects.
1b. See this script which works OK. If you enable line 2 instead of 3, 32bit integer is not working. 
```
>D
;I:zzz=19999999
zzz=19999999
>S
print S_%zzz%_
>W
nm(0 999999999 1000000 zzz "Eingabe zzz " 120)
zzz %zzz%
```
2. Yes, we both know about the existence of these {...} codes. More helpful would be an explanation. Any idea how to solve?
3. Yes, I know what you wrote. I was asking about a synchronisation to Tele period (same time).
4. A >WM section before >W section works not correct!

**Comment by gemu2015** [comment]
integer vars have only preliminary and minimal support. 
i added the nm() web cmd now.  keep in mind that integer constants must precede by the '#' char
nm(#0 #999999999 #1000000 zzz "Eingabe zzz " 120)

i added a system var 'ctper'  which gives the tele period counter. this allows to sync with tele period

the >W section must come before >WM section. 

use my fork to test the changes.  https://github.com/gemu2015/Sonoff-Tasmota

**Reply by Pewibe** [reply]
I have done your proposal by using my 2 files my_user_config.h and user_config_override.h running this script:
```
>D
I:zzz=#91000000
;x=0
>B
dp0
zzz=#92000000
>S
print S_%zzz%_
If ctper==0 
then 
print X_%ctper%_%tper%_
endif
>W
nm(#0 #999999999 #1000000 zzz "Eingabe zzz " 120)
;>WM
---
zzz %zzz%
===
```
1. Script works concerning long integers, but "If ctper==0" seems to be a problem because the following print is fired every second and not at tele period. Using >T instead of >S shows same.
2. Deleting ; in front of line 3 leads to "13:37:02.493 script init error: -1". No other variable?
3. Deleting ; in front of >WM supress the last 3 lines in script.

**Reply by gemu2015** [reply]
you have a syntax error 'if' must be lower case.

there must be at least one normal variable defined, else error -1 is thrown

section >WM is only shown by a redraw while >W is updated every 3 seconds.

**Reply by Pewibe** [reply]
1. Mea culpa!
2. You understood it vice versa. The error -1 appears if normal variable x=0 is defined. Now I figured out that the hash sign # in line 2 in front of the constant causes the problem. See the corrected script:
```
>D
I:zzz=19999999
;I:zzz=#19999999
; x=0
>S
print S_%zzz%_
if ctper==0 
then 
print X_%ctper%_%tper%_%x%_
endif
>W
nm(#0 #999999999 #1000000 zzz "Eingabe zzz " 120)
>WM
---
zzz %zzz%
===
```
3. Thanks for that hint concerning redraw (--> documentation). Still there is wrong output:
![image](https://user-images.githubusercontent.com/42893315/232809342-0317918d-ea70-4dda-838d-61cc362eae13.png)
4. And an additional question to the script master about your ScriptEditor: At each new start I have to select my font, size, etc. again. Is it possible to make this persistent?

**Reply by gemu2015** [reply]
editor font settings are currently not persistent.  will think about it when i find some time

the >WM section behaves different you must apply `<br>` for line break

**Reply by Pewibe** [reply]
Additionally to line breaks I added a dummy character at begin of each line to be wasted. Here the script and corresponding output:
![image](https://user-images.githubusercontent.com/42893315/232884556-6c48503c-2450-4af6-b78e-32f1d4c7619e.png)
Last but not least: How can I get your changes as official Tasmota version? Just wait for 12.6.0? And what about ScriptEditor "when i find some time"?
Thanks again for your quick support.

**Reply by gemu2015** [reply]
skipping a character in >WS is a bug which i did not realize before. it only appears when google charts are not defined. i always define it so i did not see this bug.  is fixed now.

Script editor has a preference file in user documents folder where you can at least edit the font type and size. (ScripEditor.plist)
i am very busy with other things so i will not change this in the near future.

i collect bugs and enhancements in my fork for some time before i do a pull request, usually several weeks

**Reply by Pewibe** [reply]
Getting finished with my script using long integer these items are still open:
5. Is it possible to produce a (formatted) string from a long integer? Something like s(2.2var)?
6. If you want to have the long integer persistent like p:var=0, what can be done? Isn't it better instead of defining by I:var, to use a recognition char, e.g. var=#123456789?

**Reply by gemu2015** [reply]
5. i don't see why this should make sense since it is an integer ? however i added this now.
6. simply  add a p before definition   p:I:var=0

**Comment by Pewibe** [comment]
5. Can you check your change concerning function s(...) or did I misunderstand something? See test script and its web output: 
```
>D
s1=""
p:I:i1=0
>B
i1=#12345678
>S
s1=s(8.2i1)
>W
nm(#0 #99999999 #1000 i1 "Eingabe " 120 3)
Label1_%8.1i1%_Dim
Label2_%s1%_Dim
```
![image](https://user-images.githubusercontent.com/42893315/233420784-6fdd9c02-0f12-4ab9-a71e-5be968a9cf30.png)

**Reply by gemu2015** [reply]
s(x) is not yet supported for integers. will check that in the next days.

**Reply by gemu2015** [reply]
fix was easy, done

**Comment by Pewibe** [comment]
You should not say "fix was easy" ;-) Now everything is perfect in my application except the possibility to assign a float f to integer i or vice versa or even calculate something like i=i+f / f=f+i. Maybe you could have a look into the code and decide what can be done.

**Reply by gemu2015** [reply]
converter functions already provided, simply use f(x) and i(x)

**Reply by Pewibe** [reply]
Great. There were so much new stuff in this thread concerning scripts that I hope the doc will be updated, to support other Tasmota users, too. I have decided to shut down the usage of rules and use scripts for all Tasmota devices. Maybe it is a good idea for supporters of Tasmota to concentrate their help for users on one alternative instead care for rules / script in parallel. Additionally with ESP32 next script appears instead of concentration.
To sum up this is what we have discussed / you have done in this thread:
1. I:vname specifies an integer 32 bit variable instead of float (limited support). Integer constants must be preceeded by '#' but not in the >D section (example I:vname=987654321. To make integers permanent, combine p: and I: to p:T:vname.
2. nm(min max step vn txt (xs) (prec)) supports both float and integer. Keep in mind that integer constants must precede by the '#' char. Example: nm(#0 #999999999 #1000000 vname "Enter vname" 120)
3. ctper as additional special variable which gives the tele period counter (0..tper). This allows to sync with tele period.
4. s(x) = explicit conversion from number x to string works now for both float and integers, but precedence by precision digits e.g. s(2.2x) is supported for float only.
5. i(x) = explicit conversion from float number x to integer.
6. f(x) = explicit conversion from integer number x to float.
7. Section >WM if used has to follow the >W section. Be aware of that line break has to be done manually, e.g. < br >. 
8. To start ScriptEditor with your personal layout you can edit the font type and size in file ScriptEditor.plist. For font names including space use parenthesis like "Courier New".

Still unanswered / a problem for me:
a. Sensors have technical names like "AM2301 Temperature" instead of user friendly. Power measurement sensor produces seldom needed lines, for example reactive power.  Using >W sections I can produce my own lookout. But I like to supress the automatically produced sensor line(s). Is there a possibility for switching on/off single lines or group of output lines?
b. In >W sections {x} codes can be used, e.g {m} generates a tab. In the doc no list of codes can be find. Who is the author Tasmota javascript who can deliver the list? Can anyboday else answer this?

**Reply by gemu2015** [reply]
@Pewibe 
according docs:  anybody can edit the docs so you are invited to contribute. just use the pencil button at the right top.
you can not suppress unwanted MQTT lines.

this is the javascript part that replaces {s}, {m} and {e}
```
".replace(/{s}/g,\"<tr><th>\")"
".replace(/{m}/g,\"</th><td style='width:20px;white-space:nowrap'>\")"  // I want a right justified column with left justified text
".replace(/{e}/g,\"</td></tr>\");"
```

**Reply by Pewibe** [reply]
a. I think you misunderstood my text. I asked how to suppress a web interface line like 
"AM2301 Taupunkt 10.0 °C" - no MQTT lines. 
b. Thanks for the code snippet. There is even a {t} and the overall head line could be "html code abbreviation". 
Your invitation concerning doc: In general I like to give something back to the community if I can. But it's not clicking the pencil for correcting a typo. Usually there is so much lack of knowledge that you are asking yourself as editor each 2nd sentence if it is correct what you are writing. I think a supervisor editor is needed as partner to ask several facts. Therefore a direct chat line with exchange and correction of text is necessary. From my feeling Tasmota today exists of a inner circle (insiders) and an outer ring of users. To contribute I think there has to be a middle ring of "co-workers" who are in special relation with a insider to process their dedicated task. 
What happens today if I would edit (and restructure) the script doc, but the author (I guess you) likes to keep his original ideas (press delete button)? All my work is wasted then. Contribution doc says: **Now you wait for one of the Tasmota admins to approve your PR (or deny 😈, no guarantees!)** Better would be that both persons first talk about general changes and get a common understanding.

**Comment by gemu2015** [comment, marked as answer]
the weboutput of a sensor driver may be suppressed by tasmota cmd "Websensor" see docs

i normally add my changes to the docs after a pr.  some people complain however the docs being not clear enough.  i take things for granted that some users don't know.

so it would be probably better if a not so experienced user edits the docs in way that he understands.
